charmhub_lp_tools/osci_sync.py

```diff
@@ -0,0 +1,247 @@
+                            accessed from the project_vars map and finally a
+                            default value if the key is not found in
+                            project_vars.
+    :returns: A new dictionary with the updated values.
+
+    """
+    new_auto_build_channel = deepcopy(auto_build_channels)
+    for lp_key, osci_key, default in auto_build_keys:
+        new_value = project_vars.get(osci_key, default)
+        if auto_build_channels.get(lp_key) != new_value:
+            logger.debug('%s channel from %s to %s',
+                         lp_key,
+                         auto_build_channels.get(lp_key),
+                         new_value)
+            new_auto_build_channel[lp_key] = new_value
+
+    return new_auto_build_channel
+
+
+def setup_parser(subparser: argparse.ArgumentParser):
+    parser = subparser.add_parser(
+        'osci-sync',
+        help='Sync the config defined in osci.yaml to Launchpad.',
+    )
+    parser.add_argument(
+        '--i-really-mean-it',
+        dest='i_really_mean_it',
+        action='store_true',
+        default=False,
+        help=('This flag must be supplied to indicate that the sync/apply '
+              'command really should be used.'),
+    )
+    parser.add_argument(
+        '--repo-dir',
+        dest='repo_dir',
+        default=os.getcwd(),
+        metavar='DIR',
+        help=('Path to the git repository where the charm is located at, by '
+              'default it uses the current working directory.'),
+    )
+    parser.set_defaults(func=main)
+    return parser
+
+
+def main(args: argparse.Namespace,
+         gc: GroupConfig,
+         ) -> None:
+    logger.setLevel(getattr(logging, args.loglevel, 'ERROR'))
+    git_repo = git.Repo(args.repo_dir, search_parent_directories=True)
+    osci = load_osci_yaml(git_repo)
+    branch_name = get_branch_name(git_repo)
+    charm_name = get_charm_name(osci)
+    project_vars = get_project_vars(osci)
+    if not list(gc.projects(select=[charm_name])):
+        logger.error("No charm '%s' found; is the name correct?", charm_name)
+        sys.exit(1)
+
+    charm_project = list(gc.projects(select=[charm_name]))[0]
+    try:
+        branch = charm_project.branches[f'refs/heads/{branch_name}']
+    except KeyError:
+        logger.error("The branch '%s' was not found in %s",
+                     branch_name, charm_project)
+        raise BranchNotFound(charm_project, branch_name)
+
+    logger.debug('channels: %s', branch['channels'])
+    for channel in branch['channels']:
+        (track, risk) = parse_channel(channel)
+        recipe_name = DEFAULT_RECIPE_FORMAT.format(
+            project=charm_project.lp_project.name,
+            branch=branch_name.replace('/', '-'),
+            track=track)
+        recipes = charm_project.lpt.get_charm_recipes(
+            charm_project.lp_team,
+            charm_project.lp_project)
+        try:
+            lp_recipe = [r for r in recipes if r.name == recipe_name][0]
```

Comment:
The `recipes = ...` may be expensive and is constant inside the for loop, so may be moved outside.  At the same time, in might be good to create a look up dictionary.  e.g.

```python
recipes = ...
lookup_map = {r.name, r for r in recipes}

# then later in the for loop:

try:
    lp_recipe = lookup_map[recipe_name]
    logger.info('Using recipe %s' ...)
except KeyError:
    logger.error('Recipe %s not found ...')
    ...
    sys.exit(2)
```

Reply:
I'm doing that here - https://github.com/freyes/charmhub-lp-tools/blob/ensure-series/charmhub_lp_tools/charm_project.py#L737-L743 - I will repeat that pattern.